src/AI/Muriel.java

```diff
@@ -0,0 +1,750 @@
+            queuedActions.offer(move);
+            break;
+          }
+        }
+        if( !queuedActions.isEmpty() ) break; // One action per invocation.
+      }
+
+      if( queuedActions.isEmpty() )
+      {
+        // Couldn't find any capture or attack actions. This unit is
+        // either a transport, or stranded on an island somewhere.
+        log(String.format("Could not find an action for %s. Waiting", unit.toStringWithLocation()));
+        queuedActions.offer(new GameAction.WaitAction(gameMap, unit, Utils.findShortestPath(unit, unit.x, unit.y, gameMap)));
+      }
+    } // ~Unit action loop
+
+    // If we don't have anything else to do, build units.
+    if( queuedActions.isEmpty() )
+    {
+      queueUnitProductionActions(gameMap);
+    }
+
+    GameAction action = queuedActions.poll();
+    log(String.format("  Action: %s", action));
+    return action;
+  }
+
+  private boolean shouldAttack(Unit unit, Unit target, GameMap gameMap)
+  {
+    // Calculate the cost of the damage we can do.
+    BattleInstance.BattleParams params = new BattleInstance.BattleParams(unit, unit.chooseWeapon(target.model, 1, true), target, gameMap.getLocation(target.x, target.y).getEnvironment());
+    double damage = params.calculateDamage();
+    UnitMatchupAndMetaInfo umami = myUnitEffectMap.get(new UnitModelPair(unit.model, target.model));
+
+    // This attack is a good idea if our cost effectiveness is in the acceptable range, or if we can at least half-kill them.
+    // The second check is needed because one glass cannon may not have a great overall ratio against another; whoever hits first wins, e.g. Mech vs Anti-Air.
+    return (umami.costEffectivenessRatio > COST_EFFECTIVENESS_THRESHOLD) || (damage > (target.getHP() / 2.0));
+  }
+
+  private void queueUnitProductionActions(GameMap gameMap)
+  {
+    log("Evaluating Production needs");
+    int budget = myCo.money;
+
+    // Get a count of enemy forces.
+    Map<UnitModel, Double> enemyUnitCounts = new HashMap<UnitModel, Double>();
+    for( Commander co : enemyCos )
+    {
+      for( Unit u : co.units )
+      {
+        if( !u.model.hasDirectFireWeapon() ) continue; // Only handle direct-fire units for now.
+        // Count how many of each model of enemy units are in play.
+        if( enemyUnitCounts.containsKey(u.model))
+        {
+          enemyUnitCounts.put(u.model, enemyUnitCounts.get(u.model) + (u.getHP() / 10) );
+        }
+        else
+        {
+          enemyUnitCounts.put(u.model, u.getHP() / 10.0 );
+        }
+      }
+    }
+
+    // Figure out what unit types we can purchase with our available properties.
+    CommanderProductionInfo CPI = new CommanderProductionInfo(myCo, gameMap);
+
+    if( CPI.availableProperties.isEmpty() )
+    {
+      log("No properties available to build.");
+      return;
+    }
+
+    // Order the enemy units by model.cost (not number), lowest first.
+    ArrayList<UnitModel> enemyModels = new ArrayList<UnitModel>();
```

Comment:
A queue would communicate intent better

Reply:
Eh, I dunno. I mean, it's ordered, because I'm sorting it, but plenty of non-queue things are sorted. I'd just as soon leave it as is.

Reply:
Fair.
I said that because you process each as you remove it, but it's not critical.